### ISSUE TITLE
white space in bold and italic causes problems

### ISSUE THREAD

**Issue by DrewNeon** [showdownjs/showdown#907]
`**bold**` is surely converted to `<strong>bold</strong>`. However, it won't convert if there's a white space at the beginning or the end within **, i.e. `** bold**`, `**bold **`, and `** bold **`. This problem also applys to `*italic*` and `***bold&italic***`, but not to  `~~strikethrough~~` or ` ``code`` `. Is it a bug?

**Comment by bandantonio** [comment]
No, it's not. According to [John Gruber's specification](https://daringfireball.net/projects/markdown/syntax#em):

> But if you surround an * or _ with spaces, it’ll be treated as a literal asterisk or underscore.

It means that `** bold**`, `**bold **`,  `** bold **` are considered syntax errors, so it’s ok that Showdown doesn’t convert them. The same applies to both `italic`, `bold + italic` and other combinations of text emphasis.

The cases with `code` and `strikethrough` are different because they are expected to show the content inside "as is" so it is expected that white spaces are taken into account.

**Comment by tivie** [comment]
## emphasis and strong

The behavior for emphasis and strong (italic and bold), is correct. 

In the [original spec](https://daringfireball.net/projects/markdown/syntax#em), it says:

> But if you surround an * or _ with spaces, it’ll be treated as a literal asterisk or underscore.

The behavior is also consistent with the [commonmark spec](https://spec.commonmark.org/0.30/#emphasis-and-strong-emphasis) of which GFM is based.

## Code

The behavior for inline code is also correct, according to the [original spec](https://daringfireball.net/projects/markdown/syntax#code) and [commonmark spec](https://spec.commonmark.org/0.30/#code-spans). See the examples in both links.

## Strikethrough

Strikethrough is a GFM specific extension. According to the [GFM spec](https://github.github.com/gfm/#strikethrough-extension-):

> Strikethrough text is any text wrapped in two tildes (~).

So, according to the written spec, `~~this ~~` should be parsed as strikethrough. However, if you test it in github, ~~this ~~ is not being parsed as strikethrough. Regardless, this is a different issue altogether.

**Comment by DrewNeon** [comment]
Thanks for the prompt responses! Yes, I didn't dig into the specs, and `**`, `~~`, ```` are so similar to the users, making me feel that they "should" behave the same. Of couse, a beginning or ending white space makes no sense, you can't really bold or italic a white space. It's just that the inconsistency in the spes costs extra efferts in developing WYSIWYG a markdown editor, so does that bold and italic share the same symbol. It's also a pity that markdown has no syntax for underline nor center alignment, given the handy HTML tags `<u></u>` and `<center></center>`. This might not be a big deal for tech docs, but could hinder markdown's usability to broader scenarios.

BTW, since @tivie mentioned github, is there a standalone javascript version of github's markdown editor available?

**Comment by tivie** [comment]
Not that I'm aware of. Github editor uses redcarpet, which is written in Ruby.

**Comment by DrewNeon** [comment]
After futher playing around, I found something confusing. 

`**abc**de**` is surely parsed to `<strong>abc</strong>de**`. As previously discussed in this thread, when I add a white space before `abc`, the first `**` is not considered as a symbol for bold syntax, i.e. `** abc**de**` is parsed to `** abc<strong>de</strong>`. This is already understood. Weird thing is when I add a white space after `abc`, I'd expect the same as a white space before `abc`, but it's parsed to `**abc *<em>de</em>*`. I just wonder why.

The above abstract example may seem meaningless. I managed to produce a real scenario.

`"Never trouble ** **trouble** ** till ** **trouble** ** troubles you." The bold words within **, i.e. **trouble**, are nouns.`

The following line is directly parsed by github's editor, which is the supposed outcome.
> "Never trouble ** **trouble** ** till ** **trouble** ** troubles you." The bold words within **, i.e. **trouble**, are nouns.

However, showdown parses it to
> "Never trouble ** **trouble** ** till ** **trouble** ** troubles you." The bold words within **, i.e. `*<em>trouble</em>*`, are nouns.

Please note the code part above and do copy the full example code to showdown and see.

**Comment by tivie** [comment]
yeah, It's a bug. These nested bold/italic cases are a pain

**Comment by DrewNeon** [comment]
Indeed, bold and italic sharing the same symbol is really a pain in the ass! I think the bug is related to the white space directly before the closing `**`, please try `**abc **de**`.  It is parsed correctly if you escape the first `**`.

Moreover, I also tried some extreme instances, six * before and after a word, i.e. `******abc******`, is parsed correctly, but seven or eight * not.

**Comment by tivie** [comment]
yeah.
Usually, if you want max portability, it is good practice to mix and match `_` and `*`.

For instance, in your case:

```
"Never trouble ** __trouble__ ** till ** __trouble__ ** troubles you." The bold words within **, i.e. __trouble__, are nouns.
```

> "Never trouble ** __trouble__ ** till ** __trouble__ ** troubles you." The bold words within **, i.e. __trouble__, are nouns.

[see demo](http://demo.showdownjs.com/#!/%22Never%20trouble%20**%20__trouble__%20**%20till%20**%20__trouble__%20**%20troubles%20you.%22%20The%20bold%20words%20within%20**%2C%20i.e.%20__trouble__%2C%20are%20nouns.)

**Comment by DrewNeon** [comment]
Exactly! It's much easier and clearer to use different symbols for each and every syntax, `**_bold&italic_**` is definately better than `***bold&italic***`.  If you are not bothered with that the same `*` symbol must be parsed differently according to the repeating times, this bug is automatically evitable. Don't understad why markdown allows such a confusion after developing so many years.
